pennylane/resource/resource.py

```diff
@@ -167,3 +170,31 @@ def __eq__(self, other: "Resources") -> bool:
                 self.gate_types == other.gate_types,
             ),
         )
+
+
+def count_resources(tape: QuantumTape, shots: int) -> Resources:
+    """Given a quantum workflow (tape) and number of samples, return the resources used.
+
+    Args:
+        tape (.QuantumTape): The quantum workflow for which we count resources
+        shots (int): The number of samples or shots to execute
+
+    Returns:
+        (.Resources): The total resources used in the workflow
+    """
+    num_wires = len(tape.wires)
+    depth = tape.graph.get_depth()
+
+    num_gates = 0
+    gate_types = defaultdict(int)
+    for op in tape.operations:
+        if hasattr(op, "resources"):
+            op_resource = op.resources()
+            for d in op_resource.gate_types:
+                gate_types[d] += op_resource.gate_types[d]
+            num_gates += sum(op_resource.gate_types.values())
+        else:
+            gate_types[op.name] += 1
+            num_gates += 1
+
```

Comment:
We can have an extra logic to optionally expand the operation, if requested, similar to the `expansion_strategy = device`  in `qml.specs`. Something like (the names are arbitrary):

```
if expansion_strategy = device
    decomp = op.decomposition()
    for d in decomp:
        # gate_types
        resource.gate_types[d.name] += 1
        ...
```

Reply:
We could add the logic you suggest, but I think the result will be inconsistent since not all PennyLane operations by default decompose to the same basis gate set. I actually think gate decomposition is a non-trivial thing to support correctly/consistently. 

It would be more modular if we kept this function to ONLY count resources, and then direct users to other tape transforms they can use to do the expansion / decomposition as they require.

Reply:
What is currently returned by `count_resources` when `expansion_strategy` is used in `qml.specs`. Is the `gate_types` consistent with what `qml.specs` returns?

Reply:
Do we need them to be consistent? We could probably pull out the expansion logic in `qml.specs` and pass the expanded tape to this function already.

Reply:
`We could probably pull out the expansion logic in qml.specs`

Yes, makes sense.